lightning/src/routing/scoring.rs

```diff
@@ -460,25 +463,425 @@ impl<T: Time> Readable for ChannelFailure<T> {
+	fn max_liquidity_msat(&self) -> u64 {
+		self.capacity_msat
+			.checked_sub(self.decayed_offset_msat(*self.max_liquidity_offset_msat))
+			.unwrap_or(0)
+	}
+
+	fn decayed_offset_msat(&self, offset_msat: u64) -> u64 {
+		self.now.duration_since(*self.last_updated).as_secs()
+			.checked_div(self.half_life.as_secs())
+			.and_then(|decays| offset_msat.checked_shr(decays as u32))
+			.unwrap_or(0)
+	}
+}
+
+impl<L: DerefMut<Target = u64>, T: Time, U: DerefMut<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat`.
+	fn failed_at_channel(&mut self, amount_msat: u64) {
+		if amount_msat < self.max_liquidity_msat() {
+			self.set_max_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat` downstream.
+	fn failed_downstream(&mut self, amount_msat: u64) {
+		if amount_msat > self.min_liquidity_msat() {
+			self.set_min_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when successfully routing `amount_msat`.
+	fn successful(&mut self, amount_msat: u64) {
+		let max_liquidity_msat = self.max_liquidity_msat().checked_sub(amount_msat).unwrap_or(0);
+		self.set_max_liquidity_msat(max_liquidity_msat);
+	}
+
+	/// Adjusts the lower bound of the channel liquidity balance in this direction.
+	fn set_min_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.min_liquidity_offset_msat = amount_msat;
+		*self.max_liquidity_offset_msat = if amount_msat > self.max_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.max_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
+	}
+
+	/// Adjusts the upper bound of the channel liquidity balance in this direction.
+	fn set_max_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.max_liquidity_offset_msat = self.capacity_msat.checked_sub(amount_msat).unwrap_or(0);
+		*self.min_liquidity_offset_msat = if amount_msat < self.min_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
+	}
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> Score for ProbabilisticScorerUsingTime<G, T> {
+	fn channel_penalty_msat(
+		&self, short_channel_id: u64, amount_msat: u64, capacity_msat: u64, source: &NodeId,
+		target: &NodeId
+	) -> u64 {
+		if *source == self.node_id || *target == self.node_id {
+			return 0;
+		}
+
+		let liquidity_penalty_multiplier_msat = self.params.liquidity_penalty_multiplier_msat;
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let success_probability = self.channel_liquidities
+			.get(&short_channel_id)
+			.unwrap_or(&ChannelLiquidity::new())
+			.as_directed(source, target, capacity_msat, liquidity_offset_half_life)
+			.success_probability(amount_msat);
+		if success_probability == 0.0 {
+			u64::max_value()
```

Comment:
I'm a bit worried about how slow our decay is in this context. I think a half-life of an hour for our min-max is probably roughly right, maybe even too fast, but given we refuse to ever try a channel if it is outside our excepted range, this may cause some issues. Specifically, I think we should be willing to retry a channel if its our only path (or only reasonably-priced path) even if it failed a half hour ago. I'm not sure quite what the right thing is, but maybe bounding the probability as at-least-1% or something like that would be an easy fix.

Reply:
agree. I would also start testing with a much faster decay of the half life.

Reply:
> bounding the probability as at-least-1%

By this do you mean have `success_probability` should never return `0.0` even if `amount_msat > max_liquidity_msat`?

Reply:
Note that lower bounding the success probability to `0.01` effectively limits the penalty to 2x `liquidity_penalty_multiplier_msat` since `-log(0.01)` is `2`.

Reply:
Yea, I think that's reasonable. I guess we could do `0.01` or `0.05` if we find that's not aggressive enough, but I have no strong intuition one way or the other.

Reply:
Let's keep it at `0.01` for now. 2x is a bit more easier for users to reason about than 1.3x when adjusting the parameters.

Reply:
It's up in `success_probability` but was squashed into this commit here:

https://github.com/lightningdevkit/rust-lightning/pull/1227/commits/560d4f0e8350ce3a5d44174edb8d38f54e571a11#diff-b00a115e9303b98772e06346a86a0312aa9790d78a73904eb6a20f938cd6da0fR598

It's a bit cleaner conditioning on the success probability here in `channel_penalty_msat` since the `success_probability` calculation can vary later without effecting the cost/penalty calculation done here, which may be generalized as a cost later including fees.

Otherwise, we'd have to pass in `liquidity_penalty_multiplier_msat` to `success_probability` and do some special casing around the zero probability. Doesn't seem as maintainable as the code progresses to include fees. Always could use an enum again for the probability if disabling clippy here is a problem.

Reply:
> It's a bit cleaner conditioning on the success probability here in channel_penalty_msat since the success_probability calculation can vary later without effecting the cost/penalty calculation done here, which may be generalized as a cost later including fees.

Ah that's....awkward. I'm not a huge fan of dead code, tbh, let alone dead branches in super hot inner loops. I wouldn't be surprised if LLVM optimizes this down with a `success_probability` inline, but I'd rather not rely on that. Given `success_probability` is an internal function and defined only a handful of lines up, would it suffice for maintainability to add a comment here?

I'll go benchmark with and without the checks just in case, though I doubt it'll be super visible.

Reply:
FWIW dropping all of the branches (even hitting the log10 call for 0) is very marginally faster in the benchmarks than either having both branches or only the `==0` branch.

Reply:
Ended up adding a debug asset and unconditionally calling `log10`. Similar performance gain to approximating `log10`, but I guess there's at less complexity here.